From be3e3770d6b5e3253d6cae23c53f9c12de8eef5f Mon Sep 17 00:00:00 2001
From: Josua Mayer <josua@solid-run.com>
Date: Thu, 24 Oct 2024 16:17:49 +0200
Subject: [PATCH 05/16] armv8: lx2160a: enable workaround for SPI erratum
 A-050752

When RCW is loaded from SDHC1, chip-selects signals for SPI3 are always
low and not usable.

Implement workaround for erratum A-050752 by clearing rcw source values
in dynamic configuration register and masking HRESET_B.

Signed-off-by: Josua Mayer <josua@solid-run.com>
---
 arch/arm/cpu/armv8/fsl-layerscape/Kconfig      |  4 ++++
 arch/arm/cpu/armv8/fsl-layerscape/cpu.c        |  3 +++
 arch/arm/cpu/armv8/fsl-layerscape/soc.c        | 18 ++++++++++++++++++
 arch/arm/include/asm/arch-fsl-layerscape/soc.h |  4 ++++
 4 files changed, 29 insertions(+)

diff --git a/arch/arm/cpu/armv8/fsl-layerscape/Kconfig b/arch/arm/cpu/armv8/fsl-layerscape/Kconfig
index 27509fcca2c..6937d9f5b92 100644
--- a/arch/arm/cpu/armv8/fsl-layerscape/Kconfig
+++ b/arch/arm/cpu/armv8/fsl-layerscape/Kconfig
@@ -308,6 +308,7 @@ config ARCH_LX2160A
 	select SYS_FSL_EC2
 	select SYS_FSL_ERRATUM_A050204
 	select SYS_FSL_ERRATUM_A011334
+	select SYS_FSL_ERRATUM_A050752
 	select SYS_FSL_ESDHC_UNRELIABLE_PULSE_DETECTION_WORKAROUND
 	select SYS_FSL_HAS_RGMII
 	select SYS_FSL_HAS_SEC
@@ -667,6 +668,9 @@ config SYS_FSL_ERRATUM_A009660
 config SYS_FSL_ERRATUM_A050382
 	bool
 
+config SYS_FSL_ERRATUM_A050752
+	bool
+
 config SYS_FSL_HAS_RGMII
 	bool
 	depends on SYS_FSL_EC1 || SYS_FSL_EC2
diff --git a/arch/arm/cpu/armv8/fsl-layerscape/cpu.c b/arch/arm/cpu/armv8/fsl-layerscape/cpu.c
index 12d31184ad9..ac39f11e3b3 100644
--- a/arch/arm/cpu/armv8/fsl-layerscape/cpu.c
+++ b/arch/arm/cpu/armv8/fsl-layerscape/cpu.c
@@ -1128,6 +1128,9 @@ int arch_early_init_r(void)
 #endif
 #if defined(CONFIG_SYS_FSL_ERRATUM_A009942) && defined(CONFIG_SYS_FSL_DDR)
 	erratum_a009942_check_cpo();
+#endif
+#ifdef CONFIG_SYS_FSL_ERRATUM_A050752
+	erratum_a050752();
 #endif
 	if (check_psci()) {
 		debug("PSCI: PSCI does not exist.\n");
diff --git a/arch/arm/cpu/armv8/fsl-layerscape/soc.c b/arch/arm/cpu/armv8/fsl-layerscape/soc.c
index 4c61d28c20f..d04f802967a 100644
--- a/arch/arm/cpu/armv8/fsl-layerscape/soc.c
+++ b/arch/arm/cpu/armv8/fsl-layerscape/soc.c
@@ -318,6 +318,24 @@ void erratum_a009635(void)
 	writel(val | 0x80000000, EPU_EPGCR);
 }
 #endif	/* CONFIG_SYS_FSL_ERRATUM_A009635 */
+#ifdef CONFIG_SYS_FSL_ERRATUM_A050752
+#define RESET_BASE 0x01e60000
+#define RESET_CCSR 0
+#define RESET_CCSR_HRESET_B_DIS BIT(25)
+
+void erratum_a050752(void)
+{
+	u32 __iomem *dcfg_ccsr = (u32 __iomem *)DCFG_BASE;
+	u32 __iomem *dcfg_dcsr = (u32 __iomem *)DCFG_DCSR_BASE;
+	u32 __iomem *reset_ccsr = (u32 __iomem *)RESET_BASE;
+	u32 val;
+
+	val = in_le32(dcfg_ccsr + DCFG_PORSR1 / 4);
+	val &= ~DCFG_PORSR1_RCW_SRC;
+	out_le32(dcfg_dcsr + DCFG_DCSR_PORCR1 / 4, val);
+	out_le32(reset_ccsr + RESET_CCSR / 4, RESET_CCSR_HRESET_B_DIS);
+}
+#endif	/* CONFIG_SYS_FSL_ERRATUM_A050752 */
 
 static void erratum_rcw_src(void)
 {
diff --git a/arch/arm/include/asm/arch-fsl-layerscape/soc.h b/arch/arm/include/asm/arch-fsl-layerscape/soc.h
index bd41df1be44..3e7c5b0e724 100644
--- a/arch/arm/include/asm/arch-fsl-layerscape/soc.h
+++ b/arch/arm/include/asm/arch-fsl-layerscape/soc.h
@@ -131,6 +131,10 @@ void erratum_a009635(void);
 void erratum_a010315(void);
 #endif
 
+#ifdef CONFIG_SYS_FSL_ERRATUM_A050752
+void erratum_a050752(void);
+#endif
+
 bool soc_has_dp_ddr(void);
 bool soc_has_aiop(void);
 
-- 
2.43.0

